### ISSUE TITLE
corhlpr.h requires internal DebugBreak() when compiling in Debug

### ISSUE THREAD

**Issue by k15tfu** [dotnet/coreclr#20833]
Hi!

I played with Profiler API (clr-samples) on Linux and got undefined-reference-to-DebugBreak when tried to compile these samples in Debug. More precisely, `COR_ILMETHOD_SECT_EH_CLAUSE_SMALL::SetTryOffset` (for example) uses `_ASSERTE` which requires the above function if `_DEBUG` is defined.

I was linking with prebuilt coreclr (`dotnet-runtime-2.1.4-linux-x64.tar.gz`) which was built in Release, obviously.  So is there the right way to build your app in Debug when using official builds?

**Comment by sywhang** [comment]
>when tried to compile these samples in Debug

Could you please specify what you mean by this?  Are you trying to compile the ReJITEnterLeaveHooks profiler? If so, you can just build it with Debug mode if you follow the steps in README. For that particular sample you need to clone and build the CoreCLR repo so that it can find the necessary header / other binary files that are used. 

From the README in that sample repo:
```
### Environment

``build.sh`` expects the following environment variables to be setup; default values are shown below.

bash
export CORECLR_PATH=~/coreclr # default
export BuildOS=Linux # Linux(default), MacOSX
export BuildArch=x64 # x64 (default)
export BuildType=Debug # Debug(default), Release
export Output=CorProfiler.so # default

``CORECLR_PATH`` is the path to your cloned and successfully built CoreCLR repository.

``BuildOS``, ``BuildArch`` and ``BuildType`` must match how you built the CoreCLR repository, so the header files and other artifacts needed for compilation are found.
```

**Comment by k15tfu** [comment]
@sywhang I want to build ReJITEnterLeaveHooks profiler in Debug mode and test it using the prebuilt coreclr (for example `dotnet-runtime-2.1.4-linux-x64.tar.gz`).  I don't want to build my own CoreCLR, especially in Debug mode (because usually this isn't required), and link with it because in this case I have to distribute them together.

**Comment by k15tfu** [comment]
Friendly ping.

**Comment by sywhang** [comment]
@k15tfu Sorry I didn't realize that you pinged me on this earlier. Must've slipped through my inbox. Feel free to ping me again with an "@" if I don't respond timely as I may miss emails :-) 

I believe for that particular profiler, what it does is that it uses some definitions from files that are inside the ```src\inc``` directory in the CoreCLR repo (it doesn't use it to run with it). So you don't have to build your own CoreCLR (you just need to clone it). You also don't have to run debug build of the runtime with it. 

Alternatively, you can bring over the missing definitions in this repo from CoreCLR repo so that it becomes more self-contained but that seems like. 

To summarize, that `CORECLR_PATH` environment variable is there for compilation purpose. You should be able to run it on retail builds without any issue.

**Comment by k15tfu** [comment]
> Feel free to ping me again with an "@" if I don't respond timely

okay deal =)

Yes, I use them. But this issue is about `_ASSERT` in `COR_ILMETHOD_SECT_EH_CLAUSE_SMALL` class, because every time I use one of its method (`SetTryOffset`, for example) in Debug I get undefined references to `GetCurrentProcessId` &  `DebugBreak`.

What we can do with it?  For now I manually commented them out =(

**Comment by k15tfu** [comment]
@sywhang Friendly ping

**Comment by sywhang** [comment]
Ah, I see what you mean... Thanks for reporting this! We should be able to just remove that when we compile on Linux. I'll put up a PR that fixes this shortly.

**Comment by sywhang** [comment]
Actually, would you be willing to share what you're trying to compile? If I try to compile the ReJITEnterLeaveHooks profiler here https://github.com/Microsoft/clr-samples/tree/master/ProfilingAPI/ReJITEnterLeaveHooks under Debug mode, but couldn't repro your issue.

**Comment by k15tfu** [comment]
@sywhang Oh, sorry for that =) Didn't notice your message.

Just add `-D_DEBUG` in CXX_FLAGS and you will get
```
/tmp/ILRewriter-c8cbd5.o: In function `ILRewriter::ImportIL(unsigned char const*)':
ILRewriter.cpp:(.text._ZN10ILRewriter8ImportILEPKh[_ZN10ILRewriter8ImportILEPKh]+0x12c): undefined reference to `GetCurrentProcessId'
ILRewriter.cpp:(.text._ZN10ILRewriter8ImportILEPKh[_ZN10ILRewriter8ImportILEPKh]+0x163): undefined reference to `DebugBreak'
ILRewriter.cpp:(.text._ZN10ILRewriter8ImportILEPKh[_ZN10ILRewriter8ImportILEPKh]+0x1c5): undefined reference to `GetCurrentProcessId'
ILRewriter.cpp:(.text._ZN10ILRewriter8ImportILEPKh[_ZN10ILRewriter8ImportILEPKh]+0x202): undefined reference to `DebugBreak'
and so on..
```

**Comment by sywhang** [comment]
Sorry for the delay @k15tfu - I'd been heads down with some other high-pri work items. I submitted a PR that should address this. Thanks for reporting this issue!
